### ISSUE TITLE
WebAssembly binary built targeting the WASI interface cannot read from standard input

### ISSUE THREAD

**Issue by flavio** [tinygo-org/tinygo#1505]
Reading from standard input from a Go binary compiled for the WASI interface is not working.
It seems the WASI backend of TinyGo doesn't have `fd_read` implemented.

# How to reproduce the issue

Create a simple program named `stdin-test.go`:

```go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	fmt.Println("say something")
	msg, err := reader.ReadString('\n')
	if err != nil {
		fmt.Printf("I got an error: %v\n", err)
	} else {
		fmt.Printf("This is what I got: |%s|\n", msg)
	}
}
```

Build the program:

```bash
$ tinygo build -o stdin-test.wasm -wasm-abi=generic -target=wasi stdin-test.go
```

Execute the program:

```bash
$ wasmtime run stdin-test.wasm 
say something
I got an error: read /dev/stdin: operation not supported
```

# Environment details

* tinygo version 0.16.0 linux/amd64 (using go version go1.15.2 and LLVM version 10.0.1)
* wasmtime 0.21.0

**Comment by flavio** [comment]
By looking at the code I see `fd_read` support is not yet done (there's even a `TODO`). I tried to add it, but failed... I would like to help fixing this issue, but I would need some generic guidance :heart:

**Comment by pojntfx** [comment]
I'm currently porting my [Berkeley sockets for WebAssembly via WebRTC](https://github.com/pojntfx/webassembly-berkeley-sockets-via-webrtc) to TinyGo and I'd greatly appreciate this as well as I'm currently using stdin/stdout in the demo programs which does not seem to be supported.

Environment:

- tinygo version 0.15.0 linux/amd64 (using go version go1.15.2 and LLVM version 10.0.1)
- Node.js v12.19.0

**Comment by trevor403** [comment]
You can get stdin using wasi if you implement the correct import bindings. I can work on getting support in xterm.js and make a PR.

**Comment by mathetake** [comment]
Hi folks, I'm the author of original WASI target PR. so please ping me when you have the PR ready for review! would love to see this PR

**Comment by aykevl** [comment]
Yes, this is something that isn't yet implemented. It will likely require some plumbing in the os package.

You should be able to call WASI functions using `//go:wasm-module` and `//export`, as long as the signatures match. This is already done in the runtime. If you need some help, feel free to ask.

**Comment by pojntfx** [comment]
While not strictly tied to WASI (`target=wasm`), here is how I've polyfilled this in the browser for now in unisockets which might be of use: https://github.com/pojntfx/unisockets/blob/103149d62ae029dfc5dfea9d8414f242ca4cb208/cmd/unisockets_runner_web/main.ts#L310-L326

**Comment by mathetake** [comment]
Update: I've started working on this issue, and the PR is almost there: https://github.com/mathetake/tinygo/compare/wasi-env-vars...mathetake:libc-wasi?expand=1 (already verified that ioutil.ReadAll works correctly).

**Comment by mathetake** [comment]
Update: I just created the PR to support basic file i/o: https://github.com/tinygo-org/tinygo/pull/1659

**Comment by deadprogram** [comment]
This was released with v0.18 so now closing. Thank you!
